File: code-example-tests/python/pymongo/examples/timeseries/aggregation_operators.py
from datetime import datetime
from pymongo import MongoClient


def load_sample_data(CONNECTION_STRING):
    try:
        client = MongoClient(CONNECTION_STRING)

        # :snippet-start: create-db
        timeseries_db = client["timeseries"]
        # :snippet-end:

        # a. create empty time series coll
        # :snippet-start: set-ts-options
        time_series_options = {
            "timeField": "date",
            "metaField": "symbol",
            "granularity": "seconds",
        }
        # :snippet-end:

        # b. create
        # :snippet-start: create-coll
        timeseries_db.create_collection("stocks", timeseries=time_series_options)
        # :snippet-end:

        # add sample docs
        # :snippet-start: add-sample-docs
        stocks_coll = timeseries_db["stocks"]

        sample_documents = [
            {
                # :snippet-start: stocks-schema
                "symbol": "MDB",
                "date": datetime(2021, 12, 18, 15, 59, 0, 0),
                "close": 252.47,
                "volume": 55046.00,
            }# :snippet-end:
            ,
            {
                "symbol": "MDB",
                "date": datetime(2021, 12, 18, 15, 58, 0, 0),
                "close": 252.93,
                "volume": 44042.00,
            },
            {
                "symbol": "APPL",
                "date": datetime(2021, 12, 18, 15, 57, 0, 0),
                "close": 253.61,
                "volume": 40182.00,
            },
            {
                "symbol": "GOOG",
                "date": datetime(2021, 12, 18, 15, 56, 0, 0),
                "close": 253.63,
                "volume": 27890.00,
            },
            {
                "symbol": "GOOG",
                "date": datetime(2021, 12, 18, 15, 55, 0, 0),
                "close": 254.03,
                "volume": 40270.00,
            },
        ]

        stocks_coll.insert_many(sample_documents)
        # :snippet-end:

        client.close()

    except Exception as e:
        raise Exception("Error while loading sample data: ", e)


def pipeline_average_price(CONNECTION_STRING):
    try:
        client = MongoClient(CONNECTION_STRING)
        timeseries_db = client["timeseries"]
        stocks_coll = timeseries_db["stocks"]

        # :snippet-start: pipeline-avg-monthly-price
        pipeline_results = stocks_coll.aggregate(
            [
                {
                    "$group": {
                        "_id": {
                            "firstDayOfMonth": {
                                "$dateTrunc": {
                                    "date": "$date",
                                    "unit": "month"
                                }
                            },
                            "symbol": "$symbol"
                        },
                        "avgMonthClose": {"$avg": "$close"}
                    }
                }
            ]
        )
        # :snippet-end:

        docs = []
        for doc in pipeline_results:
            docs.append(doc)
        client.close()
        return docs

    except Exception as e:
        raise Exception("Error while performing metafield query: ", e)


def pipeline_rolling_average(CONNECTION_STRING):
    try:
        client = MongoClient(CONNECTION_STRING)

        timeseries_db = client["timeseries"]
        stocks_coll = timeseries_db["stocks"]

        # :snippet-start: pipeline-rolling-average
        pipeline_results = stocks_coll.aggregate(
            [
                {
                    "$setWindowFields": {
                        "partitionBy": "$symbol",
                        "sortBy": {"date": 1},
                        "output": {
                            "averageMonthClosingPrice": {
                                "$avg": "$close",
                                "window": {"range": [-1, 0], "unit": "month"}
                            }
                        }
                    }
                }
            ]
        )
        # :snippet-end:

        docs = []
        for doc in pipeline_results:
            docs.append(doc)
        client.close()
        return docs

    except Exception as e:
        raise Exception("The following error occurred: ", e)

File: code-example-tests/java/driver-sync/src/main/java/timeseries/AggregationOperators.java
package timeseries;

import com.mongodb.client.*;
import com.mongodb.client.model.*;
import org.bson.Document;
import com.mongodb.client.MongoClient;
import com.mongodb.client.MongoClients;
import com.mongodb.client.MongoCollection;
import com.mongodb.client.MongoDatabase;
import com.mongodb.client.FindIterable;
import com.mongodb.client.model.CreateCollectionOptions;
import com.mongodb.client.model.TimeSeriesGranularity;
import com.mongodb.client.model.TimeSeriesOptions;
import org.bson.Document;
import org.bson.conversions.Bson;

import java.time.Instant;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Date;
import java.util.List;

public class AggregationOperators {
    private String uri = System.getenv("CONNECTION_STRING");
    private MongoClient mongoClient;
    private MongoCollection<Document> stocks;
    private String collectionName = "dowJonesTickerData";
    private MongoDatabase db;

    public void loadSampleData() {
        mongoClient = MongoClients.create(uri);
        db = mongoClient.getDatabase("timeseries_db");

        TimeSeriesOptions timeSeriesOptions = new TimeSeriesOptions("date")
                .metaField("symbol")
                .granularity(TimeSeriesGranularity.SECONDS);

        CreateCollectionOptions options = new CreateCollectionOptions()
                .timeSeriesOptions(timeSeriesOptions);
        db.createCollection(collectionName, options);

        stocks = db.getCollection(collectionName);
        stocks.deleteMany(Filters.empty());

        stocks.insertMany(
                Arrays.asList(
                        // :snippet-start: stocks-schema
                        new Document("symbol", "MDB")
                                .append("date", Date.from(Instant.parse("2021-12-18T15:59:00Z")))
                                .append("close", 252.47)
                                .append("volume", 55046.0),
                        // :snippet-end:
                        new Document("symbol", "MDB")
                                .append("date", Date.from(Instant.parse("2021-12-18T15:58:00Z")))
                                .append("close", 252.93)
                                .append("volume", 44042.0),

                        new Document("symbol", "APPL")
                                .append("date", Date.from(Instant.parse("2021-12-18T15:57:00Z")))
                                .append("close", 253.61)
                                .append("volume", 40182.0),

                        new Document("symbol", "GOOG")
                                .append("date", Date.from(Instant.parse("2021-12-18T15:56:00Z")))
                                .append("close", 253.63)
                                .append("volume", 27890.0),

                        new Document("symbol", "GOOG")
                                .append("date", Date.from(Instant.parse("2021-12-18T15:55:00Z")))
                                .append("close", 254.03)
                                .append("volume", 40270.0)
                )
        );
    }

    public List<Document> runAveragePricePipeline() {
        loadSampleData();
        // :snippet-start: pipeline-avg-monthly-price
        // Create the compound _id with $dateTrunc and symbol
        Document groupId = new Document("firstDayOfMonth",
                new Document("$dateTrunc",
                        new Document("date", "$date")
                                .append("unit", "month")))
                .append("symbol", "$symbol");

        // Create the $group stage
        Bson groupStage = Aggregates.group(
                groupId,
                Accumulators.avg("avgMonthClose", "$close")
        );

        // Run the aggregation
        List<Document> result = stocks.aggregate(List.of(groupStage))
                .into(new ArrayList<>());
        // :snippet-end:
        return result;
    }

    public List<Document> runRollingAveragePipeline() {
        loadSampleData();
        // :snippet-start: pipeline-rolling-average
        // Use $setWindowFields to specify the 30-day window
        List<Document> pipeline = List.of(
                new Document("$setWindowFields",
                        new Document("partitionBy", new Document("symbol", "$symbol"))
                                .append("sortBy", new Document("date", 1))
                                .append("output", new Document("averageMonthClosingPrice",
                                        new Document("$avg", "$close")
                                                .append("window", new Document("range", Arrays.asList(-1, "current"))
                                                        .append("unit", "month")))))
        );
        // Run the aggregation
        List<Document> result = stocks.aggregate(pipeline)
                .into(new ArrayList<>());
        // :snippet-end:
        return result;
    }

    public void Cleanup(){
        db.drop();
    }

    public String getCollectionName() {
        return collectionName;
    }

    public void setCollectionName(String collectionName) {
        this.collectionName = collectionName;
    }
}
